textattack/search_methods/improved_genetic_algorithm.py

```diff
@@ -0,0 +1,257 @@
+        If the child fails to satisfy the constraits, we re-try crossover for a fix number of times,
+        before taking one of the parents at random as the resulting child.
+        Args:
+            pop_member1 (PopulationMember): The first population member.
+            pop_member2 (PopulationMember): The second population member.
+        Returns:
+            A population member containing the crossover.
+        """
+        x1_text = pop_member1.attacked_text
+        x2_text = pop_member2.attacked_text
+
+        num_tries = 0
+        passed_constraints = False
+        while num_tries < self.max_crossover_retries + 1:
+            new_replaced_times = np.copy(pop_member1.replaced_times)
+            crossover_point = np.random.randint(0, len(new_replaced_times))
+            indices_to_replace = []
+            words_to_replace = []
+            for i in range(crossover_point, len(new_replaced_times)):
+                indices_to_replace.append(i)
+                words_to_replace.append(x2_text.words[i])
+                new_replaced_times[i] = pop_member2.replaced_times[i]
+            new_text = x1_text.replace_words_at_indices(
+                indices_to_replace, words_to_replace
+            )
+            if "last_transformation" in x1_text.attack_attrs:
+                new_text.attack_attrs["last_transformation"] = x1_text.attack_attrs[
+                    "last_transformation"
+                ]
+                filtered = self.filter_transformations(
+                    [new_text], x1_text, original_text=self.original_attacked_text
+                )
+            elif "last_transformation" in x2_text.attack_attrs:
+                new_text.attack_attrs["last_transformation"] = x2_text.attack_attrs[
+                    "last_transformation"
+                ]
+                filtered = self.filter_transformations(
+                    [new_text], x1_text, original_text=self.original_attacked_text
+                )
+            else:
+                # In this case, neither x_1 nor x_2 has been transformed,
+                # meaning that new_text == original_text
+                filtered = [new_text]
+
+            if filtered:
+                new_text = filtered[0]
+                passed_constraints = True
+                break
+
+            num_tries += 1
+
+        if not passed_constraints:
+            # If we cannot find a child that passes the constraints,
+            # we just randomly pick one of the parents to be the child for the next iteration.
+            new_text = (
+                pop_member1.attacked_text
+                if np.random.uniform() < 0.5
+                else pop_member2.attacked_text
+            )
+
+        new_results, self._search_over = self.get_goal_results([new_text])
+
+        return PopulationMember(new_text, new_replaced_times, new_results[0])
+
+    def _perform_search(self, initial_result):
+        self.original_attacked_text = initial_result.attacked_text
```

Comment:
Two things. There should be `self._search_over = False` before every search just so that we don't happen to prematurely end search if `self._search_over=True` from previous search. Also, do you mind removing `self.original_attacked_text` and switch to passing it as explicit function argument? This way, we minimize the "stateful-ness" of the search method.